overlord/fdestate/backend/reseal.go

```diff
@@ -114,49 +127,89 @@ func ResealKeyForBootChains(method device.SealingMethod, rootdir string, params
+func resealRunObjectKeys(updateState StateUpdater, pbc boot.PredictableBootChains, authKeyFile string, roleToBlName map[bootloader.Role]string) error {
 	// get model parameters from bootchains
 	modelParams, err := boot.SealKeyModelParams(pbc, roleToBlName)
 	if err != nil {
 		return fmt.Errorf("cannot prepare for key resealing: %v", err)
 	}
 
+	numModels := len(modelParams)
+	if numModels < 1 {
+		return fmt.Errorf("at least one set of model-specific parameters is required")
+	}
+
+	pcrProfile, err := secbootBuildPCRProtectionProfile(modelParams)
+	if err != nil {
+		return err
+	}
+
+	var models []secboot.ModelForSealing
+	for _, m := range modelParams {
+		models = append(models, m.Model)
+	}
+
 	// list all the key files to reseal
 	keyFiles := []string{device.DataSealedKeyUnder(boot.InitramfsBootEncryptionKeyDir)}
 
 	resealKeyParams := &secboot.ResealKeysParams{
-		ModelParams:          modelParams,
+		PCRProfile:           pcrProfile,
 		KeyFiles:             keyFiles,
 		TPMPolicyAuthKeyFile: authKeyFile,
 	}
 	if err := secbootResealKeys(resealKeyParams); err != nil {
 		return fmt.Errorf("cannot reseal the encryption key: %v", err)
 	}
 
+	// TODO: use constants for "run+recover" and "all"
+	if err := updateState("run+recover", "all", []string{"run", "recover"}, models, pcrProfile); err != nil {
+		return err
+	}
+
 	return nil
 }
 
-func resealFallbackObjectKeys(pbc boot.PredictableBootChains, authKeyFile string, roleToBlName map[bootloader.Role]string) error {
+func resealFallbackObjectKeys(updateState StateUpdater, pbc boot.PredictableBootChains, authKeyFile string, roleToBlName map[bootloader.Role]string) error {
 	// get model parameters from bootchains
 	modelParams, err := boot.SealKeyModelParams(pbc, roleToBlName)
 	if err != nil {
 		return fmt.Errorf("cannot prepare for fallback key resealing: %v", err)
 	}
 
+	numModels := len(modelParams)
+	if numModels < 1 {
+		return fmt.Errorf("at least one set of model-specific parameters is required")
+	}
+
+	pcrProfile, err := secbootBuildPCRProtectionProfile(modelParams)
+	if err != nil {
+		return err
+	}
+
+	var models []secboot.ModelForSealing
+	for _, m := range modelParams {
+		models = append(models, m.Model)
+	}
+
 	// list all the key files to reseal
 	keyFiles := []string{
 		device.FallbackDataSealedKeyUnder(boot.InitramfsSeedEncryptionKeyDir),
 		device.FallbackSaveSealedKeyUnder(boot.InitramfsSeedEncryptionKeyDir),
 	}
 
 	resealKeyParams := &secboot.ResealKeysParams{
-		ModelParams:          modelParams,
+		PCRProfile:           pcrProfile,
 		KeyFiles:             keyFiles,
 		TPMPolicyAuthKeyFile: authKeyFile,
 	}
 	if err := secbootResealKeys(resealKeyParams); err != nil {
 		return fmt.Errorf("cannot reseal the fallback encryption keys: %v", err)
 	}
 
+	// TODO: use constants for "recover" (the first parameter) and "system-save"
+	if err := updateState("recover", "system-save", []string{"recover", "factory-reset"}, models, pcrProfile); err != nil {
```

Comment:
In the spec, the "recover" key slot role has different parameters depending on the storage container. The one for the "system-data" container only needs the "recover" boot mode. The one for the "system-save" container is the only one that requires both "recover" and "factory-reset".

At the moment, it looks like we're missing a few bits that I assume will come in future PRs:
- Support for the "run" key slot role, which permits "run" mode for key slots on any storage container, although I think that's ok for now as long as the state is pre-populated with the name of the key slot role as I mentioned in another comment.
- Support for the "recover" key slot role on the "system-data" container, which excludes "factory-reset".

I assume no other properties have changed? Eg, the "run+recover" key slot role includes all recovery systems and the try model. The "recover" key slot roles only include tested recovery systems and no try model. The "run" keyslot role will include no recovery systems but will include the try model.